File: build-system.py
#!/usr/bin/env python3
"""
Ticru.io Build System
Automates the build, test, and deployment process
"""

import os
import sys
import subprocess
import argparse
from pathlib import Path
from datetime import datetime

class Colors:
    HEADER = '\033[95m'
    OKBLUE = '\033[94m'
    OKCYAN = '\033[96m'
    OKGREEN = '\033[92m'
    WARNING = '\033[93m'
    FAIL = '\033[91m'
    ENDC = '\033[0m'
    BOLD = '\033[1m'

def print_header(message):
    print(f"\n{Colors.HEADER}{Colors.BOLD}{'='*60}{Colors.ENDC}")
    print(f"{Colors.HEADER}{Colors.BOLD}{message}{Colors.ENDC}")
    print(f"{Colors.HEADER}{Colors.BOLD}{'='*60}{Colors.ENDC}\n")

def print_success(message):
    print(f"{Colors.OKGREEN}✓ {message}{Colors.ENDC}")

def print_error(message):
    print(f"{Colors.FAIL}✗ {message}{Colors.ENDC}")

def print_info(message):
    print(f"{Colors.OKCYAN}ℹ {message}{Colors.ENDC}")

def run_command(command, check=True):
    """Run a shell command and return the result"""
    try:
        result = subprocess.run(
            command,
            shell=True,
            check=check,
            capture_output=True,
            text=True
        )
        return result
    except subprocess.CalledProcessError as e:
        print_error(f"Command failed: {command}")
        print_error(f"Error: {e.stderr}")
        if check:
            sys.exit(1)
        return e

def check_dependencies():
    """Check if required dependencies are installed"""
    print_header("Checking Dependencies")
    
    dependencies = {
        'node': 'node --version',
        'npm': 'npm --version',
        'python3': 'python3 --version',
        'git': 'git --version'
    }
    
    all_ok = True
    for name, command in dependencies.items():
        result = run_command(command, check=False)
        if result.returncode == 0:
            version = result.stdout.strip()
            print_success(f"{name}: {version}")
        else:
            print_error(f"{name}: Not found")
            all_ok = False
    
    return all_ok

def install_npm_dependencies():
    """Install npm dependencies"""
    print_header("Installing NPM Dependencies")
    print_info("Running npm install...")
    result = run_command("npm install")
    if result.returncode == 0:
        print_success("NPM dependencies installed successfully")
    return result.returncode == 0

def install_python_dependencies():
    """Install Python dependencies"""
    print_header("Installing Python Dependencies")
    if os.path.exists("requirements.txt"):
        print_info("Running pip install...")
        result = run_command("pip install -r requirements.txt")
        if result.returncode == 0:
            print_success("Python dependencies installed successfully")
        return result.returncode == 0
    else:
        print_info("No requirements.txt found, skipping")
        return True

def lint_code():
    """Run linting checks"""
    print_header("Linting Code")
    
    # TypeScript/JavaScript linting
    if os.path.exists("package.json"):
        print_info("Running ESLint...")
        result = run_command("npm run lint", check=False)
        if result.returncode == 0:
            print_success("JavaScript/TypeScript linting passed")
        else:
            print_error("JavaScript/TypeScript linting failed")
            return False
    
    return True

def type_check():
    """Run TypeScript type checking"""
    print_header("Type Checking")
    
    if os.path.exists("tsconfig.json"):
        print_info("Running TypeScript type check...")
        result = run_command("npm run type-check", check=False)
        if result.returncode == 0:
            print_success("Type checking passed")
        else:
            print_error("Type checking failed")
            return False
    
    return True

def build_frontend():
    """Build the frontend application"""
    print_header("Building Frontend")
    print_info("Running production build...")
    result = run_command("npm run build")
    if result.returncode == 0:
        print_success("Frontend built successfully")
        
        # Check build output
        dist_path = Path("dist")
        if dist_path.exists():
            files = list(dist_path.rglob("*"))
            print_success(f"Build output: {len(files)} files in dist/")
        
        return True
    return False

def test_code():
    """Run tests"""
    print_header("Running Tests")
    print_info("Tests not configured yet")
    return True

def generate_build_info():
    """Generate build information file"""
    print_header("Generating Build Info")
    
    build_info = {
        'version': '1.0.0',
        'build_date': datetime.now().isoformat(),
        'git_commit': run_command('git rev-parse HEAD', check=False).stdout.strip(),
        'git_branch': run_command('git rev-parse --abbrev-ref HEAD', check=False).stdout.strip()
    }
    
    info_path = Path('dist/build-info.json')
    if info_path.parent.exists():
        import json
        with open(info_path, 'w') as f:
            json.dump(build_info, f, indent=2)
        print_success(f"Build info generated: {info_path}")
    
    return True

def clean_build():
    """Clean build artifacts"""
    print_header("Cleaning Build Artifacts")
    
    paths_to_clean = ['dist', 'build', 'node_modules/.cache']
    
    for path in paths_to_clean:
        if os.path.exists(path):
            print_info(f"Removing {path}...")
            run_command(f"rm -rf {path}")
            print_success(f"Removed {path}")
    
    return True

def main():
    parser = argparse.ArgumentParser(description='Ticru.io Build System')
    parser.add_argument('--clean', action='store_true', help='Clean build artifacts')
    parser.add_argument('--install', action='store_true', help='Install dependencies')
    parser.add_argument('--lint', action='store_true', help='Run linting')
    parser.add_argument('--type-check', action='store_true', help='Run type checking')
    parser.add_argument('--build', action='store_true', help='Build the application')
    parser.add_argument('--test', action='store_true', help='Run tests')
    parser.add_argument('--all', action='store_true', help='Run complete build pipeline')
    
    args = parser.parse_args()
    
    # If no specific action, run all
    if not any([args.clean, args.install, args.lint, args.type_check, args.build, args.test, args.all]):
        args.all = True
    
    start_time = datetime.now()
    print_header(f"Ticru.io Build System - {start_time.strftime('%Y-%m-%d %H:%M:%S')}")
    
    success = True
    
    # Check dependencies first
    if not check_dependencies():
        print_error("Missing required dependencies")
        sys.exit(1)
    
    # Execute requested actions
    if args.clean or args.all:
        success = success and clean_build()
    
    if args.install or args.all:
        success = success and install_npm_dependencies()
        success = success and install_python_dependencies()
    
    if args.lint or args.all:
        success = success and lint_code()
    
    if args.type_check or args.all:
        success = success and type_check()
    
    if args.build or args.all:
        success = success and build_frontend()
        if success:
            success = success and generate_build_info()
    
    if args.test or args.all:
        success = success and test_code()
    
    # Summary
    end_time = datetime.now()
    duration = (end_time - start_time).total_seconds()
    
    print_header("Build Summary")
    print(f"Duration: {duration:.2f} seconds")
    
    if success:
        print_success("Build completed successfully!")
        sys.exit(0)
    else:
        print_error("Build failed!")
        sys.exit(1)

if __name__ == "__main__":
    main()

File: build-system.ts
#!/usr/bin/env node

/**
 * Ticru.io Build System
 * Automates the build, test, and deployment process
 */

import { execSync, ExecException } from 'child_process';
import { existsSync, writeFileSync, rmSync } from 'fs';
import { Command } from 'commander';

// ANSI color codes
const Colors = {
  HEADER: '\x1b[95m',
  OKBLUE: '\x1b[94m',
  OKCYAN: '\x1b[96m',
  OKGREEN: '\x1b[92m',
  WARNING: '\x1b[93m',
  FAIL: '\x1b[91m',
  ENDC: '\x1b[0m',
  BOLD: '\x1b[1m',
};

function printHeader(message: string): void {
  console.log(`\n${Colors.HEADER}${Colors.BOLD}${'='.repeat(60)}${Colors.ENDC}`);
  console.log(`${Colors.HEADER}${Colors.BOLD}${message}${Colors.ENDC}`);
  console.log(`${Colors.HEADER}${Colors.BOLD}${'='.repeat(60)}${Colors.ENDC}\n`);
}

function printSuccess(message: string): void {
  console.log(`${Colors.OKGREEN}✓ ${message}${Colors.ENDC}`);
}

function printError(message: string): void {
  console.log(`${Colors.FAIL}✗ ${message}${Colors.ENDC}`);
}

function printInfo(message: string): void {
  console.log(`${Colors.OKCYAN}ℹ ${message}${Colors.ENDC}`);
}

function runCommand(command: string, check: boolean = true): boolean {
  try {
    execSync(command, { stdio: 'inherit' });
    return true;
  } catch (error) {
    const e = error as ExecException;
    printError(`Command failed: ${command}`);
    if (e.stderr) {
      printError(`Error: ${e.stderr}`);
    }
    if (check) {
      process.exit(1);
    }
    return false;
  }
}

function checkDependencies(): boolean {
  printHeader('Checking Dependencies');

  const dependencies = {
    node: 'node --version',
    npm: 'npm --version',
    git: 'git --version',
  };

  let allOk = true;
  for (const [name, command] of Object.entries(dependencies)) {
    try {
      const version = execSync(command, { encoding: 'utf-8' }).trim();
      printSuccess(`${name}: ${version}`);
    } catch (error) {
      printError(`${name}: Not found`);
      allOk = false;
    }
  }

  return allOk;
}

function installNpmDependencies(): boolean {
  printHeader('Installing NPM Dependencies');
  printInfo('Running npm install...');
  const success = runCommand('npm install');
  if (success) {
    printSuccess('NPM dependencies installed successfully');
  }
  return success;
}

function lintCode(): boolean {
  printHeader('Linting Code');

  if (existsSync('package.json')) {
    printInfo('Running ESLint...');
    const success = runCommand('npm run lint', false);
    if (success) {
      printSuccess('JavaScript/TypeScript linting passed');
    } else {
      printError('JavaScript/TypeScript linting failed');
      return false;
    }
  }

  return true;
}

function typeCheck(): boolean {
  printHeader('Type Checking');

  if (existsSync('tsconfig.json')) {
    printInfo('Running TypeScript type check...');
    const success = runCommand('npm run type-check', false);
    if (success) {
      printSuccess('Type checking passed');
    } else {
      printError('Type checking failed');
      return false;
    }
  }

  return true;
}

function buildFrontend(): boolean {
  printHeader('Building Frontend');
  printInfo('Running production build...');
  const success = runCommand('npm run build');
  if (success) {
    printSuccess('Frontend built successfully');

    // Check build output
    if (existsSync('dist')) {
      const files = require('fs').readdirSync('dist', { recursive: true });
      printSuccess(`Build output: ${files.length} files in dist/`);
    }

    return true;
  }
  return false;
}

function buildApi(): boolean {
  printHeader('Building API Server');
  printInfo('Compiling TypeScript API...');
  const success = runCommand('npm run build:api');
  if (success) {
    printSuccess('API server built successfully');
    return true;
  }
  return false;
}

function testCode(): boolean {
  printHeader('Running Tests');
  printInfo('Tests not configured yet');
  return true;
}

function generateBuildInfo(): boolean {
  printHeader('Generating Build Info');

  try {
    const gitCommit = execSync('git rev-parse HEAD', { encoding: 'utf-8' }).trim();
    const gitBranch = execSync('git rev-parse --abbrev-ref HEAD', {
      encoding: 'utf-8',
    }).trim();

    const buildInfo = {
      version: '1.0.0',
      build_date: new Date().toISOString(),
      git_commit: gitCommit,
      git_branch: gitBranch,
    };

    if (existsSync('dist')) {
      writeFileSync('dist/build-info.json', JSON.stringify(buildInfo, null, 2));
      printSuccess('Build info generated: dist/build-info.json');
    }

    return true;
  } catch (error) {
    printError('Failed to generate build info');
    return false;
  }
}

function cleanBuild(): boolean {
  printHeader('Cleaning Build Artifacts');

  const pathsToClean = ['dist', 'build', 'node_modules/.cache'];

  for (const path of pathsToClean) {
    if (existsSync(path)) {
      printInfo(`Removing ${path}...`);
      rmSync(path, { recursive: true, force: true });
      printSuccess(`Removed ${path}`);
    }
  }

  return true;
}

async function main() {
  const program = new Command();

  program
    .name('build-system')
    .description('Ticru.io Build System')
    .option('--clean', 'Clean build artifacts')
    .option('--install', 'Install dependencies')
    .option('--lint', 'Run linting')
    .option('--type-check', 'Run type checking')
    .option('--build', 'Build the application')
    .option('--build-api', 'Build the API server')
    .option('--test', 'Run tests')
    .option('--all', 'Run complete build pipeline');

  program.parse();

  const options = program.opts();

  // If no specific action, run all
  if (
    !options.clean &&
    !options.install &&
    !options.lint &&
    !options.typeCheck &&
    !options.build &&
    !options.buildApi &&
    !options.test &&
    !options.all
  ) {
    options.all = true;
  }

  const startTime = Date.now();
  printHeader(`Ticru.io Build System - ${new Date().toLocaleString()}`);

  let success = true;

  // Check dependencies first
  if (!checkDependencies()) {
    printError('Missing required dependencies');
    process.exit(1);
  }

  // Execute requested actions
  if (options.clean || options.all) {
    success = success && cleanBuild();
  }

  if (options.install || options.all) {
    success = success && installNpmDependencies();
  }

  if (options.lint || options.all) {
    success = success && lintCode();
  }

  if (options.typeCheck || options.all) {
    success = success && typeCheck();
  }

  if (options.build || options.all) {
    success = success && buildFrontend();
    if (success) {
      success = success && generateBuildInfo();
    }
  }

  if (options.buildApi || options.all) {
    success = success && buildApi();
  }

  if (options.test || options.all) {
    success = success && testCode();
  }

  // Summary
  const endTime = Date.now();
  const duration = (endTime - startTime) / 1000;

  printHeader('Build Summary');
  console.log(`Duration: ${duration.toFixed(2)} seconds`);

  if (success) {
    printSuccess('Build completed successfully!');
    process.exit(0);
  } else {
    printError('Build failed!');
    process.exit(1);
  }
}

main().catch((error) => {
  printError(`Build system error: ${error.message}`);
  process.exit(1);
});
